### PR TITLE
Add maintenance jobs feature

### PR DESCRIPTION
**Pull request by gingters**

### CODE REVIEW

**Comment by gingters** [comment]
Closes #278

File: src/Thinktecture.Relay.Server.Abstractions/Maintenance/IMaintenanceJob.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+	/// <summary>
+	/// An implementation of a job that will be instanciated and called from a background task in a configured maintenance interval.
+	/// </summary>
+	public interface IMaintenanceJob
+	{
+		/// <summary>
+		/// Called when the maintenance interval
+		/// </summary>
+		/// <param name="stoppingToken">Indicates that the maintenance loop has been aborted.</param>
+		/// <returns>An awaitable task.</returns>
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server.Abstractions/Maintenance/IMaintenanceJob.cs:15]:
```suggestion
		/// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
```

File: src/Thinktecture.Relay.Server/Maintenance/MaintenanceJobRunner.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+					_logger.LogTrace("Running maintenance job {MaintenanceJob}.", job.GetType().FullName);
+					await job.DoMaintenanceAsync(stoppingToken);
+				}
+				catch (TaskCanceledException)
+				{
+					// Ignore this, as this will be thrown when the service shuts down gracefully
+				}
+				catch (Exception ex)
+				{
+					_logger.LogError(ex, "An error occured while running maintenance job {MaintenanceJob}.", job.GetType().FullName);
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server/Maintenance/MaintenanceJobRunner.cs:75]:
```suggestion
					_logger.LogError(ex, "An error occured while running maintenance job {MaintenanceJob}", job.GetType().FullName);
```

File: src/Thinktecture.Relay.Server/Maintenance/StatisticsCleanupJob.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		public StatisticsCleanupJob(IStatisticsRepository statisticsRepository, IOptions<StatisticsOptions> statisticsOptions)
+		{
+			_statisticsRepository = statisticsRepository ?? throw new ArgumentNullException(nameof(statisticsRepository));
+			_statisticsOptions = statisticsOptions.Value ?? throw new ArgumentNullException(nameof(statisticsOptions));
+		}
+
+		/// <inheritdoc />
+		public async Task DoMaintenanceAsync(CancellationToken stoppingToken)
+		{
+			if (stoppingToken.IsCancellationRequested)
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server/Maintenance/StatisticsCleanupJob.cs:29]: Why is this necessary? Isn't it better to forward the token to the repository calls instead?

File: src/docker/Thinktecture.Relay.Server.Docker/Startup.cs [hunk 4]
```diff
@@ -36,7 +37,11 @@ public void ConfigureServices(IServiceCollection services)
 			services.AddRelayServer()
 				.AddRabbitMqRouting(options => Configuration.GetSection("RabbitMq").Bind(options))
 				.AddSignalRConnectorTransport()
-				.AddFileBodyStore(options => Configuration.GetSection("BodyStore").Bind(options));
+				.AddFileBodyStore(options => Configuration.GetSection("BodyStore").Bind(options))
+				.AddMaintenanceJobs(options => Configuration.GetSection("Maintenance").Bind(options))
+				;
```
Comment (thomashilzendegen) [inline src/docker/Thinktecture.Relay.Server.Docker/Startup.cs:42]:
```suggestion
				.AddMaintenanceJobs(options => Configuration.GetSection("Maintenance").Bind(options));
```

File: src/docker/Thinktecture.Relay.Server.Docker/appsettings.json [hunk 5]
```diff
@@ -22,5 +22,12 @@
     "Limits": {
       "MaxRequestBodySize": 115343360
     }
+  },
+  "Maintenance": {
+    "RunInterval": "00:10:00" // run ever 10 minutes, default: 15 Minutes
+  },
+  "Statistics": {
+    "EntryMaxAge": "1:00:00", // one hour, default: 15 Minutes
```
Comment (thomashilzendegen) [inline src/docker/Thinktecture.Relay.Server.Docker/appsettings.json:30]:
```suggestion
    "EntryMaxAge": "01:00:00", // one hour, default: 15 Minutes
```

File: src/Thinktecture.Relay.Server.Persistence.EntityFrameworkCore/StatisticsRepository.cs [hunk 6]
```diff
@@ -82,9 +82,9 @@ public async Task ShutdownOriginAsync(Guid originId)
 		}
 
 		/// <inheritdoc />
-		public async Task CleanUpOriginsAsync(TimeSpan oldestToKeep)
+		public async Task CleanUpOriginsAsync(TimeSpan maxAge)
 		{
-			var filterTime = DateTime.UtcNow - oldestToKeep;
+			var filterTime = DateTime.UtcNow - maxAge;
 
 			_logger.LogDebug("Cleaning up statistics storage: Deleting all origins that have not been updated since {OldestOriginToKeep}.", filterTime);
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server.Persistence.EntityFrameworkCore/StatisticsRepository.cs:89]:
```suggestion
			_logger.LogDebug("Cleaning up statistics storage by deleting all origins that have not been seen since {OriginLastSeen}.", filterTime);
```

File: src/Thinktecture.Relay.Server.Persistence.EntityFrameworkCore/StatisticsRepository.cs [hunk 7]
```diff
@@ -165,9 +165,9 @@ public async Task CloseConnectionAsync(string connectionId)
 		}
 
 		/// <inheritdoc />
-		public async Task CleanUpConnectionsAsync(TimeSpan oldestToKeep)
+		public async Task CleanUpConnectionsAsync(TimeSpan maxAge)
 		{
-			var filterTime = DateTime.UtcNow - oldestToKeep;
+			var filterTime = DateTime.UtcNow - maxAge;
 
 			_logger.LogDebug("Cleaning up statistics storage: Deleting all connections that have not been updated since {OldestOriginToKeep}.", filterTime);
```
Comment (thomashilzendegen) [inline src/Thinktecture.Relay.Server.Persistence.EntityFrameworkCore/StatisticsRepository.cs:172]:
```suggestion
			_logger.LogDebug("Cleaning up statistics storage by deleting all connections that have no activity since {ConnectionLastActivity}.", filterTime);
```

File: src/Thinktecture.Relay.Server/Maintenance/StatisticsCleanupJob.cs (hunk 3 above)
Comment (gingters) [reply src/Thinktecture.Relay.Server/Maintenance/StatisticsCleanupJob.cs]: implemented

**Review by thomashilzendegen — APPROVED** [review]
